File: TPs/TP01/Q12-83/PalindromoRec.c
/**
Refaça em C a questão anterior.
*/
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

#define NUMLINHA    1000
#define TAMPALAVRA  1000



/**
*   Encontrar fim
*   @param  String a testar
*   @return true -> fim     encontrado
*           false-> fim nao encontrado
*/
bool isFim (char s[]){
    return (strlen(s)>=3 && s[0]=='F' && s[1]=='I' && s[2]=='M');
}

/**
*   Testa se Palindromo
*   @param  s       -  string a testar
*   @param  start   -  primeira posicao
*   @param  last    -  ultima posicao
*   @return true    - se     for palindromo
*           false   - se nao for palindromo
*/
bool isPalindromo (char *entrada, int iterador, bool resp){
    memset(entrada, 0, sizeof(entrada));
    int lastAtual = strlen(entrada);
    //
    if (iterador >= (lastAtual / 2))
        return resp;
    else
    {
        if (entrada[iterador] == entrada[(lastAtual)-iterador])
            return isPalindromo(entrada, resp, ++iterador);
        else
            return isPalindromo(entrada, false, lastAtual - 1);
    }
}

/**
Main
*/
int main (){
    //
    char entrada[10000];
    int tamanho = strlen(entrada);
    //
    bool resp = true;
    //
    do{
        memset(entrada, 0, sizeof(entrada));
        //scanf("%[^\n]%*c", entrada);
        fgets(entrada, 100, stdin);
        if(tamanho>0 && entrada[tamanho-1]=='\n'){
            entrada[tamanho-1]='\0';
        }
        if (isFim(entrada)==false){
            resp = isPalindromo(entrada, 0, true);
            if (resp==true)
                printf("NAO\n");
            else
                printf("SIM\n");
        }
    } while (isFim(entrada)==false);
    return 0;
}

File: TPs/TP01/Q11-ok/PalindromoRec.java
class PalindromoRec {
    /**
    Encontrar fim
    @param  s       - String a testar
    @return true    - fim     encontrado
    *       false   - fim nao encontrado
    */
    public static boolean isFim (String s){
        return (s.length()==3 && s.charAt(0)=='F'&& s.charAt(1)=='I'&& s.charAt(2)=='M');
    }
    /**
    Testa se Palindromo
    @param  s       -  string a testar
    @param  start   -  primeira posicao
    @param  last    -  ultima posicao
    @return true    - se     for palindromo
    *       false   - se nao for palindromo
    */
    public static boolean isPalindromo (String s, int start, int last){
        //se fim da string
        //  fim atual da string
        int lastAtual = s.length();
        //  funcao logaritmica - 1
        if(((lastAtual/2)-1)==last){
            //fim da string
            return true;
        }
        //testar se nao palindromo
        if(s.charAt(start) == s.charAt(last)){
            //  teste
            //MyIO.println(s.charAt(start)+" - "+s.charAt(last));
            return isPalindromo(s, start+1, last-1);
        }
        //nao palindromo
        return false;
    }
    /**
    Main
    */
    public static void main (String[] args){
        //Variaveis de entrada
        String[] entrada = new String[1000];
        int numEntrada=0;
        //Leitura da entrada padrao
        // while not 'FIM'
        do{
            entrada[numEntrada]=MyIO.readLine();
            //If para impedir e entrada da palavra fim
            if (isFim(entrada[numEntrada])==false)
                //Se palindromo print SIM/NAO
                if(isPalindromo(entrada[numEntrada], 0, entrada[numEntrada].length()-1)==true)
                    MyIO.println("SIM");
                else
                    MyIO.println("NAO");
        } while (isFim(entrada[numEntrada++])==false);
        numEntrada--;   //Desconsidera a linha da palavra FIM
    }
}
